CRM/Case/BAO/Case.php

```diff
@@ -1217,24 +1209,52 @@ public static function getCaseActivity($caseID, &$params, $contactID, $context =
       // if there are file attachments we will return how many and, if only one, add a link to it
       if (!empty($dao->attachment_ids)) {
         $attachmentIDs = array_unique(explode(',', $dao->attachment_ids));
-        $caseActivity['no_attachments'] = count($attachmentIDs);
+        $caseActivities[$caseActivityId]['no_attachments'] = count($attachmentIDs);
         $url .= implode(' ', CRM_Core_BAO_File::paperIconAttachment('civicrm_activity', $caseActivityId));
       }
 
-      $caseActivity['links'] = $url;
-
-      array_push($caseActivities, $caseActivity);
+      $caseActivities[$caseActivityId]['links'] = $url;
     }
     $dao->free();
 
     $caseActivitiesDT = array();
-    $caseActivitiesDT['data'] = $caseActivities;
+    $caseActivitiesDT['data'] = array_values($caseActivities);
     $caseActivitiesDT['recordsTotal'] = $caseCount;
     $caseActivitiesDT['recordsFiltered'] = $caseCount;
 
     return $caseActivitiesDT;
   }
 
+  /**
+   * Helper function to generate a formatted contact link/name for display in the Case activities tab
+   *
+   * @param $contactId
+   * @param $contactName
+   *
+   * @return string
+   */
+  private static function formatContactLink($contactId, $contactName) {
+    if (!isset(Civi::$statics[__CLASS__]['contactViewUrl'])) {
+      Civi::$statics[__CLASS__]['contactViewUrl'] = CRM_Utils_System::url("civicrm/contact/view", "reset=1&cid=", FALSE, NULL, FALSE);
+    }
+    $contactViewUrl = Civi::$statics[__CLASS__]['contactViewUrl'];
+
+    if (!isset(Civi::$statics[__CLASS__]['hasViewContact'])) {
+      Civi::$statics[__CLASS__]['hasViewContact'] = CRM_Core_Permission::giveMeAllACLs();
+    }
+    $hasViewContact = Civi::$statics[__CLASS__]['hasViewContact'];
```

Comment:
Instead of checking all acls, you can specifically check permission for a given contact using the function `CRM_Contact_BAO_Contact_Permission::allow()`.

Reply:
@colemanw I was trying to make minimal changes to the code, and it was using the `giveMeAllACLs()` function before this PR - I just moved it to it's own function and added a static cache around it.
It looks like `CRM_Contact_BAO_Contact_Permission::allow()` is basically a drop-in replacement and the static cache would still be useful.  If you agree, then I'll change the function.

Reply:
I think let's drop the static cache per my comment above and swap to the more accurate permission checking function. Thanks :)